Using the limited Python API in astropy

When building a Python package with C extensions, it is possible to build it against only the limited Python C API (more details at https://docs.python.org/3/c-api/stable.html), which is an API that does not change across minor versions of Python. This makes it possible to build a wheel on e.g. 3.11 and have it also work on 3.12 and 3.13. As an example, Sunpy uses this, and they only have to provide a few wheels, one per platform/architecture:

https://pypi.org/project/sunpy/6.1.1/

Extension-helpers will soon include a change that makes it possible to easily opt in to using this by just adding a single configuration item in ``pyproject.toml``, namely:

```
[tool.distutils.bdist_wheel]
py-limited-api = "cp311"
```

This also requires using Cython 3.1.

Currently this does not work out of the box for astropy core, and there are a few errors we need to resolve:

**column_mixins.pyx**

It looks like Cython support may not be complete here, we should try and see if we can make a minimal example of the failure here and report it upstream.

```
x86_64-linux-gnu-gcc -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DPy_LIMITED_API=0x030B0000 -DNPY_TARGET_VERSION=NPY_1_23_API_VERSION -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -I/home/tom/python/dev/lib/python3.11/site-packages/numpy/_core/include -I/home/tom/python/dev/include -I/usr/include/python3.11 -c astropy/table/_column_mixins.c -o build/temp.linux-x86_64-cpython-311/astropy/table/_column_mixins.o
astropy/table/_column_mixins.c: In function ‘__pyx_f_7astropy_5table_14_column_mixins_column_getitem’:
astropy/table/_column_mixins.c:5217:80: error: invalid use of incomplete typedef ‘PyTypeObject’ {aka ‘struct _typeobject’}
 5217 |   __pyx_t_1 = ((PyTypeObject *)__pyx_mstate_global->__pyx_ptype_5numpy_ndarray)->tp_as_mapping->mp_subscript(__pyx_v_self, __pyx_v_item); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 74, __pyx_L1_error)
      |                                                                                ^~
error: command '/usr/lib/ccache/x86_64-linux-gnu-gcc' failed with exit code 1
```

**storage size unknown**

There are several similar errors:


```
astropy/utils/xml/src/iterparse.c: At top level:
astropy/utils/xml/src/iterparse.c:1105:21: error: storage size of ‘IterParserType’ isn’t known
 1105 | static PyTypeObject IterParserType =
      |                     ^~~~~~~~~~~~~~
error: command '/usr/lib/ccache/x86_64-linux-gnu-gcc' failed with exit code 1
```

```
astropy/wcs/src/astropy_wcs.c:753:21: error: storage size of ‘WcsType’ isn’t known
  753 | static PyTypeObject WcsType = {
      |                     ^~~~~~~
astropy/wcs/src/astropy_wcs.c:753:21: error: storage size of ‘WcsType’ isn’t known
error: command '/usr/lib/ccache/x86_64-linux-gnu-gcc' failed with exit code 1
```

```
astropy/cosmology/_src/signature_deprecations.c: At top level:
astropy/cosmology/_src/signature_deprecations.c:200:21: error: storage size of ‘DeprKwsWrap’ isn’t known
  200 | static PyTypeObject DeprKwsWrap = {
      |                     ^~~~~~~~~~~
cc1: all warnings being treated as errors
error: command '/usr/lib/ccache/x86_64-linux-gnu-gcc' failed with exit code 1
```

Aside from these issues, everything builds fine, so it's not completely impossible to get there!

Need to investigate further what this is due to (similar to astropy.utils error above)

I wanted to take a shot a fixing the existing errors, and to my surprise, I don't seem to get the same ones your reported.
I do get the 3 following errors (alongside 10 warnings that I'll skip for brevity):

```
(...)
astropy/wcs/src/astropy_wcs.c:94:16: error: incomplete definition of type 'PyTypeObject' (aka 'struct _typeobject')
   94 |   Py_TYPE(self)->tp_free((PyObject*)self);
      |   ~~~~~~~~~~~~~^
/Users/clm/.local/share/uv/python/cpython-3.11.12-macos-aarch64-none/include/python3.11/pytypedefs.h:20:16: note: forward declaration of 'struct _typeobject'
   20 | typedef struct _typeobject PyTypeObject;
      |                ^
astropy/wcs/src/astropy_wcs.c:104:20: error: incomplete definition of type 'PyTypeObject' (aka 'struct _typeobject')
  104 |   self = (Wcs*)type->tp_alloc(type, 0);
      |                ~~~~^
/Users/clm/.local/share/uv/python/cpython-3.11.12-macos-aarch64-none/include/python3.11/pytypedefs.h:20:16: note: forward declaration of 'struct _typeobject'
   20 | typedef struct _typeobject PyTypeObject;
      |                ^

(...)
astropy/wcs/src/astropy_wcs.c:753:21: error: variable has incomplete type 'PyTypeObject' (aka 'struct _typeobject')
  753 | static PyTypeObject WcsType = {
      |                     ^
/Users/clm/.local/share/uv/python/cpython-3.11.12-macos-aarch64-none/include/python3.11/pytypedefs.h:20:16: note: forward declaration of 'struct _typeobject'
   20 | typedef struct _typeobject PyTypeObject;
      |                ^
```

All these errors live in `astropy_wcs.c`, note in Cython code.

@da-woods I tried the following patch:
```patch
diff --git a/astropy/table/_column_mixins.pyx b/astropy/table/_column_mixins.pyx
index 5ab4fe66d3..f9227e533d 100644
--- a/astropy/table/_column_mixins.pyx
+++ b/astropy/table/_column_mixins.pyx
@@ -26,7 +26,7 @@ import sys
 
 import numpy as np
 cimport numpy as np
-
+from cpython.ref cimport PyTypeObject
 np.import_array()
 
 
@@ -44,8 +44,9 @@ cdef extern from "Python.h":
     ctypedef struct PyMappingMethods:
         binaryfunc mp_subscript
 
-    ctypedef struct PyTypeObject:
-        PyMappingMethods* tp_as_mapping
+    void *PyType_GetSlot(PyTypeObject *type, int slot)
+
+    int Py_tp_as_mapping
 
 
 ctypedef object (*item_getter)(object, object)
@@ -71,7 +72,7 @@ cdef inline object base_getitem(object self, object item, item_getter getitem):
 
 
 cdef inline object column_getitem(object self, object item):
-    return (<PyTypeObject *>np.ndarray).tp_as_mapping.mp_subscript(self, item)
+    return (<PyMappingMethods *>PyType_GetSlot(<PyTypeObject *>np.ndarray, Py_tp_as_mapping)).mp_subscript(self, item)
 
```

but compilation fails with

```
astropy/table/_column_mixins.c:5195:118: error: use of undeclared identifier 'Py_tp_as_mapping'
 5195 |   __pyx_t_1 = ((PyMappingMethods *)PyType_GetSlot(((PyTypeObject *)__pyx_mstate_global->__pyx_ptype_5numpy_ndarray), Py_tp_as_mapping))->mp_subscript(__pyx_v_self, __pyx_v_item); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 75, __pyx_L1_error)
      |                                                                                                                      ^
1 error generated
```

I'm guessing I'm not doing it right. What's the correct way to include a constant like `Py_tp_as_mapping` from `Python.h` in Cython ?

--- a/.github/workflows/ci_workflows.yml
+++ b/.github/workflows/ci_workflows.yml
@@ -138,6 +138,8 @@ jobs:
 
   test_limited_api_build:
     # Test to make sure that we can build astropy with the limited API
+    # This job is also important to check `build` as our build frontend
+    # see https://github.com/astropy/astropy/pull/18253
     strategy:
         fail-fast: false
         matrix:

--- a/.github/workflows/publish.yml
+++ b/.github/workflows/publish.yml
@@ -43,6 +43,7 @@ jobs:
       env: |
         CIBW_BEFORE_BUILD: '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'pip install -U --pre --extra-index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple setuptools>=77.0.0 setuptools_scm cython numpy>=0.0.dev0 extension-helpers') || '' }}'
         CIBW_BUILD_FRONTEND: '${{ ((github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'Build all wheels')) && 'pip; args: --no-build-isolation') || 'build' }}'
+        EXTENSION_HELPERS_PY_LIMITED_API: 'cp311'
 
       test_extras: test
       # FIXME: we exclude the test_data_out_of_range test since it
@@ -52,43 +53,21 @@ jobs:
       test_command: pytest -p no:warnings --astropy-header -m "not hypothesis" -k "not test_data_out_of_range and not test_set_locale and not TestQuantityTyping" --pyargs astropy
       targets: |
         # Linux wheels
-
-        - cp311-manylinux_x86_64
-        - cp312-manylinux_x86_64
-        - cp313-manylinux_x86_64
-
-        - target: cp311-manylinux_aarch64
-          runs-on: ubuntu-24.04-arm
-        - target: cp312-manylinux_aarch64
-          runs-on: ubuntu-24.04-arm
-        - target: cp313-manylinux_aarch64
+        - cp3*-manylinux_x86_64
+        - target: cp3*-manylinux_aarch64
           runs-on: ubuntu-24.04-arm
 
         # Note that following wheels are not currently tested:
-        - cp311-musllinux_x86_64
-        - cp312-musllinux_x86_64
-        - cp313-musllinux_x86_64
+        - cp3*-musllinux_x86_64
 
         # MacOS X wheels - as noted in https://github.com/astropy/astropy/pull/12379 we deliberately
         # do not build universal2 wheels.
-
-        - cp311*macosx_x86_64
-        - cp312*macosx_x86_64
-        - cp313*macosx_x86_64
-
-        - cp311*macosx_arm64
-        - cp312*macosx_arm64
-        - cp313*macosx_arm64
+        - cp3*-macosx_x86_64
+        - cp3*-macosx_arm64
 
         # Windows wheels
-
-        - cp311*win32
-        - cp312*win32
-        - cp313*win32
-
-        - cp311*win_amd64
-        - cp312*win_amd64
-        - cp313*win_amd64
+        - cp3*-win32
+        - cp3*-win_amd64
 
     secrets:
       pypi_token: ${{ secrets.pypi_token }}

--- a/astropy/time/src/parse_times.c
+++ b/astropy/time/src/parse_times.c
@@ -401,7 +401,7 @@ create_parser(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds)
     if (name == NULL) {
         name = "fast_parser";
     }
-    Py_INCREF(dt_pars);
+    Py_INCREF((PyObject*)dt_pars);
     pars_array = (PyArrayObject *)PyArray_FromAny(pars, dt_pars, 1, 1,
                      (NPY_ARRAY_CARRAY | NPY_ARRAY_ENSURECOPY), NULL);
     if (pars_array == NULL) {
@@ -432,7 +432,7 @@ create_parser(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds)
     return (PyObject *)gufunc;
 
   fail:
-    Py_XDECREF(pars_array);
+    Py_XDECREF((PyObject*)pars_array);
     Py_XDECREF((PyObject*)gufunc);
     return NULL;
 }

--- a/astropy/utils/xml/src/iterparse.c
+++ b/astropy/utils/xml/src/iterparse.c
@@ -931,7 +931,7 @@ IterParser_dealloc(IterParser* self)
         self->parser = NULL;
     }
 
-    PyTypeObject* type = Py_TYPE(self);
+    PyTypeObject* type = Py_TYPE((PyObject*)self);
     freefunc free_func = PyType_GetSlot(type, Py_tp_free);
     free_func((PyObject*)self);
 }

--- a/astropy/wcs/src/astropy_wcs.c
+++ b/astropy/wcs/src/astropy_wcs.c
@@ -64,7 +64,7 @@ Wcs_traverse(
   Py_VISIT(self->py_distortion_lookup[0]);
   Py_VISIT(self->py_distortion_lookup[1]);
   Py_VISIT(self->py_wcsprm);
-  Py_VISIT((PyObject*)Py_TYPE(self));
+  Py_VISIT((PyObject*)Py_TYPE((PyObject*)self));
 
   return 0;
 }

--- a/astropy/wcs/src/distortion_wrap.c
+++ b/astropy/wcs/src/distortion_wrap.c
@@ -17,7 +17,7 @@ PyDistLookup_traverse(
     void* arg) {
 
   Py_VISIT(self->py_data);
-  Py_VISIT((PyObject*)Py_TYPE(self));
+  Py_VISIT((PyObject*)Py_TYPE((PyObject*)self));
 
   return 0;
 }

--- a/astropy/wcs/src/str_list_proxy.c
+++ b/astropy/wcs/src/str_list_proxy.c
@@ -57,7 +57,7 @@ PyStrListProxy_traverse(
     void *arg) {
 
   Py_VISIT(self->pyobject);
-  Py_VISIT((PyObject*)Py_TYPE(self));
+  Py_VISIT((PyObject*)Py_TYPE((PyObject*)self));
   return 0;
 }
 

--- a/astropy/wcs/src/unit_list_proxy.c
+++ b/astropy/wcs/src/unit_list_proxy.c
@@ -63,7 +63,7 @@ PyUnitListProxy_traverse(
 
   Py_VISIT(self->pyobject);
   Py_VISIT(self->unit_class);
-  Py_VISIT((PyObject*)Py_TYPE(self));
+  Py_VISIT((PyObject*)Py_TYPE((PyObject*)self));
   return 0;
 }
 

--- a/astropy/wcs/src/wcslib_auxprm_wrap.c
+++ b/astropy/wcs/src/wcslib_auxprm_wrap.c
@@ -32,7 +32,7 @@ PyAuxprm_new(PyTypeObject* type, PyObject* args, PyObject* kwds) {
 static int
 PyAuxprm_traverse(PyAuxprm* self, visitproc visit, void *arg) {
   Py_VISIT(self->owner);
-  Py_VISIT((PyObject*)Py_TYPE(self));
+  Py_VISIT((PyObject*)Py_TYPE((PyObject*)self));
   return 0;
 }
 

--- a/astropy/wcs/src/wcslib_celprm_wrap.c
+++ b/astropy/wcs/src/wcslib_celprm_wrap.c
@@ -104,7 +104,7 @@ static PyObject* PyCelprm_new(PyTypeObject* type, PyObject* args, PyObject* kwds
 static int PyCelprm_traverse(PyCelprm* self, visitproc visit, void *arg)
 {
     Py_VISIT(self->owner);
-    Py_VISIT((PyObject*)Py_TYPE(self));
+    Py_VISIT((PyObject*)Py_TYPE((PyObject*)self));
     return 0;
 }
 

--- a/astropy/wcs/src/wcslib_prjprm_wrap.c
+++ b/astropy/wcs/src/wcslib_prjprm_wrap.c
@@ -101,7 +101,7 @@ static PyObject* PyPrjprm_new(PyTypeObject* type, PyObject* args, PyObject* kwds
 static int PyPrjprm_traverse(PyPrjprm* self, visitproc visit, void *arg)
 {
     Py_VISIT(self->owner);
-    Py_VISIT((PyObject*)Py_TYPE(self));
+    Py_VISIT((PyObject*)Py_TYPE((PyObject*)self));
     return 0;
 }
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -266,6 +266,7 @@ markers = [
 # We disable testing for the following wheels:
 # - MuslLinux (tests hang non-deterministically)
 test-skip = "*-musllinux_x86_64"
+environment-pass = ["EXTENSION_HELPERS_PY_LIMITED_API"]
 
 [tool.cibuildwheel.macos]
 archs = ["x86_64", "arm64"]

